src/vs/workbench/contrib/terminal/node/terminalProfiles.ts

```diff
@@ -0,0 +1,273 @@
+}
+
+interface IPotentialTerminalProfile {
+	profileName: string,
+	paths: string[],
+	args?: string[]
+}
+
+export function detectAvailableProfiles(quickLaunchOnly: boolean, logService?: ILogService, config?: ITerminalConfiguration | ITestTerminalConfig, variableResolver?: ExtHostVariableResolverService, workspaceFolder?: IWorkspaceFolder, statProvider?: IStatProvider): Promise<ITerminalProfile[]> {
+	return platform.isWindows ? detectAvailableWindowsProfiles(quickLaunchOnly, logService, config?.detectWslProfiles, config?.profiles?.windows, variableResolver, workspaceFolder, statProvider) : detectAvailableUnixProfiles();
+}
+
+async function detectAvailableWindowsProfiles(quickLaunchOnly: boolean, logService?: ILogService, detectWslProfiles?: boolean, configProfiles?: any, variableResolver?: ExtHostVariableResolverService, workspaceFolder?: IWorkspaceFolder, statProvider?: IStatProvider): Promise<ITerminalProfile[]> {
+	// Determine the correct System32 path. We want to point to Sysnative
+	// when the 32-bit version of VS Code is running on a 64-bit machine.
+	// The reason for this is because PowerShell's important PSReadline
+	// module doesn't work if this is not the case. See #27915.
+	const is32ProcessOn64Windows = process.env.hasOwnProperty('PROCESSOR_ARCHITEW6432');
+	const system32Path = `${process.env['windir']}\\${is32ProcessOn64Windows ? 'Sysnative' : 'System32'}`;
+
+	let useWSLexe = false;
+
+	if (getWindowsBuildNumber() >= 16299) {
+		useWSLexe = true;
+	}
+
+	let expectedProfiles: IPotentialTerminalProfile[] = [
+		{
+			profileName: 'Command Prompt',
+			paths: [`${system32Path}\\cmd.exe`]
+		},
+		{
+			profileName: 'Git Bash',
+			paths: [
+				`${process.env['ProgramW6432']}\\Git\\bin\\bash.exe`,
+				`${process.env['ProgramW6432']}\\Git\\usr\\bin\\bash.exe`,
+				`${process.env['ProgramFiles']}\\Git\\bin\\bash.exe`,
+				`${process.env['ProgramFiles']}\\Git\\usr\\bin\\bash.exe`,
+				`${process.env['LocalAppData']}\\Programs\\Git\\bin\\bash.exe`,
+			],
+			args: ['--login']
+		},
+		{
+			profileName: 'Cygwin',
+			paths: [
+				`${process.env['HOMEDRIVE']}\\cygwin64\\bin\\bash.exe`,
+				`${process.env['HOMEDRIVE']}\\cygwin\\bin\\bash.exe`
+			],
+			args: ['-l']
+		},
+		... await getWslProfiles(`${system32Path}\\${useWSLexe ? 'wsl.exe' : 'bash.exe'}`, detectWslProfiles, logService),
+		... await getPowershellProfiles()
+	];
+
+	const promises: Promise<ITerminalProfile | undefined>[] = [];
+	expectedProfiles.forEach(profile => promises.push(validateProfilePaths(profile.profileName, profile.paths, statProvider, profile.args)));
+	const profiles = await Promise.all(promises);
+
+	const detectedProfiles = coalesce(profiles);
+
+	if (!quickLaunchOnly) {
+		return detectedProfiles;
+	}
+
+	let validProfiles: ITerminalProfile[] = [];
+
+	if (detectedProfiles && configProfiles) {
+		for (const [profileKey, value] of Object.entries(configProfiles)) {
+			if (value !== null) {
+				if ((value as ITerminalExecutable).path) {
```

Comment:
If value is a `ITerminalExecutable || ITerminalProfileGenerator` you can use `if ('path' in value)` which would narrow the type within the if statement since only `ITerminalExecutable` contains path

Reply:
that feels less concise to me and doesn't work when I do that

Reply:
We can keep as is to get it in, just a minor thing